contrib/go.mod

```diff
@@ -0,0 +1,23 @@
+module github.com/envoyproxy/go-control-plane/contrib
+
+go 1.20
+
+replace github.com/envoyproxy/go-control-plane/envoy => ../envoy
```

Comment:
My understanding from `go.work` is that it should no longer be needed to add those replace commands
Is that actually not the case?

Reply:
I was able to drop the replace from `./envoy` but that one still complained so I left it in.

Reply:
Maybe this is an indication that we don't have the right structure? Since the "envoy" and "contrib" protobufs are bound to the same Envoy release, does it make sense to allow the possibility for a program to import different versions of them?

How hard would it be to move the contrib protobufs to "envoy/contrib"?

Reply:
And I suppose that a similar question arises for the "./ratelimit" directory.

Reply:
OK, contrib seems weird, any maybe we want to fix it first.

The [dlb.proto](https://github.com/envoyproxy/envoy/blob/main/api/contrib/envoy/extensions/network/connection_balance/dlb/v3alpha/dlb.proto#L10) Go package option says it is in the "github.com/envoyproxy/go-control-plane/envoy" module. However, the actual source code is in "github.com/envoyproxy/go-control-plane/contrib/envoy/extensions".

Seems like the protobuf sync has been putting them in the wrong module? cc @phlax

Reply:
is that an issue with this proto ?

if i do:

```console
$ find build_go/ -name "dlb*"
build_go/contrib/envoy/extensions/network/connection_balance/dlb
build_go/contrib/envoy/extensions/network/connection_balance/dlb/v3alpha/dlb.pb.go
build_go/contrib/envoy/extensions/network/connection_balance/dlb/v3alpha/dlb.pb.validate.go
```

it appears that the path is correct going on the src path

Reply:
grepping further, none of the contrib apis appear to have contrib in their go_package path  - im not sure if i misunderstand the issue

Reply:
For example:

```
$ git grep go_package | grep vcl
api/contrib/envoy/extensions/vcl/v3alpha/vcl_socket_interface.proto:option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/vcl/v3alpha";
```

But the actual import path that works is:
```
	_ "github.com/envoyproxy/go-control-plane/contrib/envoy/extensions/vcl/v3alpha"
```

So this seems like the two repositories don't agree on where contrib ought to live?

Reply:
i suspect that when these were moved to contrib their `go_package` was never updated

Reply:
Is bazel the thing that chooses to generate them into the contrib path?

Reply:
yeah i think so - and it uses the src path i believe

Reply:
afaict none of the contrib extensions have the correct `go_package` on this basis

Reply:
```console
$ git grep go-control-plane api/contrib/ | grep go-control-plane/contrib
```

Reply:
I'm in agreement here and would like contrib to fold under `/envoy`. I didn't do much investigation but seems like its in-scope for this PR. If we are going to break that import path I would say now is the time to do it since this will be the first multi-module release.

Reply:
not sure iiuc - one thing im thinking is that if the `go_package` that envoy has listed doesnt match what it produces then we probably want to fix something there one way or other

the diff above would need to be applied in envoy, so also a bit concerned who else might be consuming it but then i think if we are fixing the package/path mismatch that is kinda ok (and perhaps suggests not so many)

Reply:
@phlax you think its a worthy issue to open in envoy? We can see what people think

Reply:
certainly if there is a mismatch i think we should

not clear what the best solution is but would totally defer to you - definitely good to get other opinions if its a breaking change tho

Reply:
> Correct me if I'm wrong the contrib folder is for things that don't see much active dev right?

contrib extensions might get active dev, depending on the instance, whether that surfaces in API changes or impl changes, I'm not sure, seems like right now API changes are less frequent

relevant: https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md#contrib-extensions

Reply:
one other (untested, so it may not work tbh) idea would be to rather than have separate `envoy` or `contrib` modules at the root or to try to reparent the `contrib` module under `envoy` would be to instead add an `api` (or some other name) module that contains the `envoy` and `contrib` packages so they can be versioned together

we could make this a purely additive change to give people time to switch over to using this new module versioning scheme/import path (i.e. keep mirroring to the existing locations in this repo for a couple releases), internally in the root module we might be able to switch over right away

i haven't thought about the specifics of modifying the import path/the `go_package` directive, but this might be a way to make that sort of breaking change easier to stomach

it adds some duplication but we can document that we will remove it soon

Reply:
Has anyone tried to use this branch in an existing project? I'm having trouble with Contour which would have to import the main module and envoy modules

Reply:
I'm getting ambiguous import issues since the new module maintains the same import path at the moment

Reply:
I did try the imports with a trivial program (that just did underscore imports from each module), and it worked for me. Testing on a real client project is much better though!

Reply:
> @jpeach Since we're creating 2 tags, one for the control server and another for everything else, would they not be versioned the same even if they're in separate folders?

When you release, I expect you would tag both of them, but projects that consume them would need to remember to update both modules together. I think that the intention of Go modules dependencies would be for the "contrib" module to require the matching version of the "envoy" module (so the release process would need to update `go.mod` in "contrib".

> contrib folder is for things that don't see much active dev right

Always having the API for contrib extensions doesn't bother me much since the IIUC, they are not enabled in the standard Envoy build.

>  instead add an api (or some other name) module that contains the envoy and contrib packages so they can be versioned together

Yes, that's what I'm proposing (by suggesting that we move the "contrib" directory inside "envoy").

Reply:
> Yes, that's what I'm proposing (by suggesting that we move the "contrib" directory inside "envoy").

I think my comment above is a slightly different, it intends to mirror the structure of the envoy `api` folder structure so `contrib` and `envoy` packages live at the same level under a new module (move/duplicate the current `envoy` and `contrib` packages into `github.com/envoyproxy/go-control-plane/api/[envoy|contrib]`)

Reply:
this branch has that set up: https://github.com/sunjayBhatia/go-control-plane/tree/multi-module-release-api-module

Reply:
> > Yes, that's what I'm proposing (by suggesting that we move the "contrib" directory inside "envoy").
> 
> I think my comment above is a slightly different, it intends to mirror the structure of the envoy api folder structure so contrib and envoy packages live at the same level under a new module

I see what you mean. That would have the same effect, but nested one directory. That sounds fine to me, and will be a pretty easy change to the envoy sync script as well. I agree with you that leaving the old packages in place makes the migration path smoother.

Reply:
heres a branch of Contour using the above branch: https://github.com/sunjayBhatia/contour/tree/use-go-control-plane-multi-module

Reply:
(the above go-control-plane branch also copies the `ratelimit` module to a package in the new `api` module)

Reply:
i think either approach will work just fine, though the upgrade/transition UX might be cleaner (no possibility for users complaining about ambiguous package paths if they are requiring mismatched versions) if the new modules have a new package path

Reply:
> (the above go-control-plane branch also copies the `ratelimit` module to a package in the new `api` module)

I'll check later, but me recollection is that the ratelimit proto it's done come from the envoy repository. So they probably need to be versioned differently and it might not make sense for them to be in the same module as the envoy api.